### PR TITLE
Conflicting pre-bookings

### LINKED ISSUE #3825: Handle conflicting pre-bookings.
We do not currently render the conflicts between overlapping pre-bookings, despite the fact that they may indeed overlap.
We had that in the previous interface.

Tasks: 
- Display conflicting pre-bookings as available when searching just with a warning (in book room dialog) that there are conflicts (similar to the "some occurrences will be skipped"): something like "conflicting pre-bookings will be rejected" if you **book** and "there are other pre-bookings overlapping" if you also just **pre-book**.
- Allow creating overlapping pre-bookings
- When one of the overlapping pre-bookings gets confirmed, cancel the remaining one.
- Display in calendar pre-booking conflicts differently ("Conflict between pre-bookings")

### PR DESCRIPTION
**Pull request by mvidalgarcia**
closes #3825
image
image

TODO:
- [x] Improve legend styling
- [x] Do more testing 
  - recurring bookings
  - pre-conflicts in different dates
  - overlapping occurrences get deleted when booking
  - etc.

### CODE REVIEW

File: indico/modules/rb/operations/conflicts.py [hunk 1]
```diff
@@ -129,3 +129,16 @@ def get_room_unbookable_hours_conflicts(candidates, occurrences):
                 obj = TempReservationOccurrence(overlap[0], overlap[1], None)
                 conflicts.add(obj)
     return conflicts, conflicting_candidates
+
+
+def get_concurrent_pre_bookings(pre_bookings, skip_conflicts_with=frozenset()):
+    concurrent_pre_bookings = set()
+    for x in pre_bookings:
+        for y in pre_bookings:
```
Comment (ThiefMaster) [inline indico/modules/rb/operations/conflicts.py]: itertools has something better for this (`product` I think); but there might be something even better (`combinations`) sine you don't need `(y, x)` after checking `(x, y)`.

File: indico/modules/rb/client/js/modules/bookRoom/reducers.js [hunk 2]
```diff
@@ -83,8 +83,13 @@ const timelineReducer = combineReducers({
                     availability: state.availability.concat(camelizeKeys(action.data.availability)),
                 };
             case actions.CREATE_BOOKING_SUCCESS: {
-                const {data: {room_id: roomId}} = action;
+                const {data: {room_id: roomId, booking: {is_accepted: isAccepted}}} = action;
                 const {roomIds, availability} = state;
+                if (!isAccepted) {
+                    const [, roomAvailability] = availability.find(([id]) => id === roomId);
+                    roomAvailability.preConflicts = roomAvailability.candidates;
```
Comment (ThiefMaster) [inline indico/modules/rb/client/js/modules/bookRoom/reducers.js]: you can't mutate existing state in a reducer...

File: indico/modules/rb/client/js/common/timeline/TimelineItem.jsx [hunk 3]
```diff
@@ -204,6 +223,38 @@ class TimelineItem extends React.Component {
                     </strong>
                 </div>
             );
+        } else if (type === 'concurrent-pre-booking') {
+            let popupMessage;
+            if (fullyOverlaps(reservations)) {
+                popupMessage = (
+                    <div>
+                        <Translate>Fully overlap between two pre-bookings:</Translate>
```
Comment (ThiefMaster) [inline indico/modules/rb/client/js/common/timeline/TimelineItem.jsx]: "Full overlap" ("Fully" would require a verb, not a noun)

**Comment by ThiefMaster** [comment]
image

The icon looks a bit odd...

**Comment by ThiefMaster** [comment]
There's another overlapping prebooking (roughly where I drew the beautiful red rectangle). Shouldn't it be shown in this view (just like the other confirmed bookings on that day)?

image

File: indico/web/client/js/utils/date.js [hunk 4]
```diff
@@ -109,3 +110,25 @@ export function initialEndTime(end) {
...
+ * Convert string date into date-agnostic datetime
+ * @param {String} dt - String datetime to transform
+ * @returns {moment.Moment} - Date-agnostic datetime
+ */
+export function getTime(dt) {
+    return moment(dt).set({year: 0, month: 0, day: 0});
+}
+
+/**
+ * Check whether there is a fully overlap between two pre-bookings
```
Comment (mic4ael) [inline indico/web/client/js/utils/date.js]: Is it correct? `there is a fully overlap`? What about `full overlap` or sth like `Check whether two pre-bookings fully overlap`

**Pull request merged**